### PR TITLE
Fix logic for `include_parent_namespaces` option

### PR DESCRIPTION
**Pull request by 304**
Based on PR #1408.

* Add specs to test different combinations of namespaces
* Add documentation for the `include_parent_namespaces` option

### CODE REVIEW

File: lib/grape/dsl/inside_route.rb [hunk 1]
```diff
@@ -29,7 +29,10 @@ module PostBeforeFilter
         def declared(params, options = {}, declared_params = nil)
           options = options.reverse_merge(include_missing: true, include_parent_namespaces: true)
 
-          declared_params ||= (!options[:include_parent_namespaces] ? route_setting(:declared_params) : (route_setting(:saved_declared_params) || [])).flatten(1) || []
+          all_declared_params = route_setting(:declared_params)
+          current_namespace_declared_params = route_setting(:saved_declared_params).last
+
+          declared_params ||= options[:include_parent_namespaces] ? all_declared_params : current_namespace_declared_params
```
Comment (304) [inline lib/grape/dsl/inside_route.rb]: I think it won't harm to provide a short explanation what is going on here.

`route_setting(:declared_params)` keep every parameter in a flat array, for example:
```
route_setting(:declared_params) # [:parent_name, :child_name, :child_age, :grandchild_name]
```
However, `route_setting(:saved_declared_params)` stores parameters in groups by namespace.
```
route_setting(:saved_declared_params) # [[:parent_name], [:child_name, :child_age], [:grandchild_name]]
```
The last element here contains only parameters that were defined in current namespace.

File: lib/grape/dsl/inside_route.rb [hunk 2]
```diff
@@ -29,7 +29,10 @@ module PostBeforeFilter
         def declared(params, options = {}, declared_params = nil)
           options = options.reverse_merge(include_missing: true, include_parent_namespaces: true)
 
-          declared_params ||= (!options[:include_parent_namespaces] ? route_setting(:declared_params) : (route_setting(:saved_declared_params) || [])).flatten(1) || []
+          all_declared_params = route_setting(:declared_params)
+          current_namespace_declared_params = route_setting(:saved_declared_params).last
```
Comment (faucct) [inline lib/grape/dsl/inside_route.rb]: Something can be broken, because you have removed that `route_setting(:saved_declared_params) || []` piece.

Comment (faucct) [inline lib/grape/dsl/inside_route.rb]: I think that just replacing `flatten(1)` with `last` is enough.

**Comment by dblock** [comment]
I like this. This just needs a CHANGELOG entry, please.

File: lib/grape/dsl/inside_route.rb (hunk 2 above)
Comment (304) [inline lib/grape/dsl/inside_route.rb]: I tried to investigate the reason why these additional checks were introduced, but it's still unclear for me (PR #739 and specifically [here](https://github.com/ruby-grape/grape/pull/739/files#diff-e9e8424a5238d48301e313d8fe285697R22)).

I also played around endpoint structure to detect the case when `flatten` was really needed. But specs were green no matter what I did. I would personally prefer to clean up the code that does not break tests. What do you think?

**Comment by 304** [comment]
@dblock, CHANGELOG updated => 8a7a24c

File: CHANGELOG.md [hunk 3]
```diff
@@ -15,6 +15,7 @@
 * [#1365](https://github.com/ruby-grape/grape/pull/1365): Fix finding exception handler in error middleware - [@ktimothy](https://github.com/ktimothy).
 * [#1380](https://github.com/ruby-grape/grape/pull/1380): Fix `allow_blank: false` for `Time` attributes with valid values causes `NoMethodError` - [@ipkes](https://github.com/ipkes).
 * [#1384](https://github.com/ruby-grape/grape/pull/1384): Fix parameter validation with an empty optional nested `Array` - [@ipkes](https://github.com/ipkes).
+* [#1415](https://github.com/ruby-grape/grape/pull/1415): Fix logic for `include_parent_namespaces` option - [@304](https://github.com/304).
```
Comment (dblock) [inline CHANGELOG.md]: It doesn't say where, it's in `declared`, maybe say "Fix `declared(params, include_parent_namespaces: true)`"?

**Comment by dblock** [comment]
Merged via https://github.com/ruby-grape/grape/commit/7d1b3e334f5d3d5b045e9a48b08a8f1ed03a9f5e.

**Pull request closed without merging**